packages/balances/src/modules/EvmErc20Module.ts

```diff
@@ -97,9 +81,47 @@ export const EvmErc20Module: NewBalanceModule<
   const { chainConnectors, chaindataProvider } = hydrate
   const chainConnector = chainConnectors.evm
   assert(chainConnector, "This module requires an evm chain connector")
+  const moduleTokens = chaindataProvider.tokensObservable.pipe(
+    map((tokens) =>
+      Object.fromEntries(
+        tokens
+          .filter((token): token is EvmErc20Token => token.type === moduleType)
+          .map((t) => [t.id, t])
+      )
+    )
+  )
+
+  const prepareFetchParameters = async (
+    addressesByToken: AddressesByToken<EvmErc20Token>,
+    allTokens: Record<string, EvmErc20Token>
+  ): Promise<EvmErc20NetworkParams> => {
+    const addressesByTokenByEvmNetwork = groupAddressesByTokenByEvmNetwork(
+      addressesByToken,
+      allTokens
+    )
+    return Object.entries(addressesByTokenByEvmNetwork).reduce(
+      (result, [evmNetworkId, addressesByToken]) => {
+        const networkParams = Object.entries(addressesByToken).flatMap(([tokenId, addresses]) => {
+          const token = allTokens[tokenId]
+
+          return addresses.map((address) => ({
+            token,
+            address,
+            evmNetworkId,
+          }))
+        })
+        result[evmNetworkId] = networkParams
+        return result
+      },
+      {} as EvmErc20NetworkParams
+    )
+  }
 
   return {
-    ...DefaultBalanceModule("evm-erc20"),
+    ...DefaultBalanceModule(moduleType),
+    get tokens() {
+      return firstValueFrom(moduleTokens)
```

Comment:
Thanks for the explanation, previously I had `chaindataProvider.tokensById()` here but changed it because of @0xKheops comment: https://github.com/TalismanSociety/talisman/pull/1380#discussion_r155518970

It may be that the whole idea of adding a `tokens` attribute to the module type was a mistake, can scrap it if it doesn't make sense

Reply:
Aye! (1) I think it makes sense to remove `tokens()` from the module type, but then it's still used inside the module so (2) I think it'd be good to replace this:
```ts
const moduleTokens = chaindataProvider.tokensObservable.pipe(
  map((tokens) =>
    Object.fromEntries(
      tokens
        .filter((token): token is EvmErc20Token => token.type === moduleType)
        .map((t) => [t.id, t])
    )
  )
)
...
// and then, where we need the tokens
await firstValueFrom(moduleTokens)
```
With this:
```ts
const getModuleTokens = async () =>
  util.itemsToMapById(
    (await chaindataProvider.tokens()).filter(
      (token): token is EvmErc20Token => token.type === moduleType
    )
  )
...
// and then, where we need the tokens
await getModuleTokens()
```

Reply:
Avoids digging into rxjs territory unless we really need to subscribe to a stream of values